addons/html_editor/static/src/editor/core/delete_plugin.js

```diff
@@ -0,0 +1,726 @@
+import {
+    DIRECTIONS,
+    boundariesOut,
+    childNodeIndex,
+    endPos,
+    leftPos,
+    nodeSize,
+    rightPos,
+} from "../utils/position";
+import { getDeepRange, setSelection } from "../utils/selection";
+import { CTGROUPS, CTYPES } from "../utils/content_types";
+import { splitTextNode } from "../utils/dom_split";
+
+const shortcuts = {
+    DELETE_BACKWARD: (e) => e.key === "Backspace" && !e.ctrlKey && !e.metaKey,
+};
+
+export class DeletePlugin extends Plugin {
+    static dependencies = ["dom"];
+    static name = "delete";
+
+    setup() {
+        this.addDomListener(this.editable, "keydown", this.handleShortcut.bind(this));
+        this.addDomListener(this.editable, "beforeinput", this.onBeforeInput.bind(this));
+    }
+    handleCommand(command, payload) {
+        switch (command) {
+            case "DELETE_BACKWARD":
+                this.deleteBackward();
+                break;
+            case "DELETE_FORWARD":
+                this.deleteForward();
+                break;
+            case "DELETE_ELEMENT_BACKWARD":
+                this.deleteElementBackward(payload);
+                break;
+            case "DELETE_ELEMENT_FORWARD":
+                this.deleteElementForward(payload);
+                break;
+            case "DELETE_RANGE":
+                this.deleteRange();
+                break;
+        }
+    }
+
+    // --------------------------------------------------------------------------
+    // commands
+    // --------------------------------------------------------------------------
+
+    deleteBackward() {
+        let selection = this.shared.getEditableSelection();
+        if (!selection) {
+            return;
+        }
+        if (handleZWS(this.editable, selection)) {
+            this.dispatch("DELETE_RANGE");
+            return;
+        }
+        selection = this.shared.getEditableSelection();
+
+        this.deleteElementBackward({
+            targetNode: selection.anchorNode,
+            targetOffset: selection.anchorOffset,
+        });
+    }
+    deleteForward() {
+        let selection = this.shared.getEditableSelection();
+        if (!selection) {
+            return;
+        }
+        if (handleZWS(this.editable, selection)) {
+            this.dispatch("DELETE_RANGE");
+            return;
+        }
+        selection = this.shared.getEditableSelection();
+
+        this.deleteElementForward({
+            targetNode: selection.anchorNode,
+            targetOffset: selection.anchorOffset,
+        });
+    }
+
+    deleteElementBackward(params) {
+        const { targetNode, targetOffset } = params;
+
+        for (const callback of this.registry.category("delete_element_backward_before").getAll()) {
+            if (callback({ ...params })) {
+                return;
+            }
+        }
+
+        const shouldStop =
+            this.deleteElementBackwardTextNode(targetNode, targetOffset) ||
```

Comment:
it is strangely elegant and scary... :)

Reply:
I agree, I don't like the names and the technique is quite odd. I'm open to suggestion 👼